cpp/src/decisiontree/decisiontree_impl.cuh

```diff
@@ -133,108 +134,75 @@ std::ostream &operator<<(std::ostream &os, const SparseTreeNode<T, L> &node) {
   return os;
 }
 
-template <typename T, typename L>
-struct Node_ID_info {
-  const SparseTreeNode<T, L> &node;
-  int unique_node_id;
-
-  Node_ID_info(const SparseTreeNode<T, L> &cfg_node, int cfg_unique_node_id)
-    : node(cfg_node), unique_node_id(cfg_unique_node_id) {}
-};
-
 template <class T, class L>
-void build_treelite_tree(TreeBuilderHandle tree_builder,
-                         DecisionTree::TreeMetaDataNode<T, L> *tree_ptr,
-                         int num_class) {
-  int node_id = 0;
-  TREELITE_CHECK(TreeliteTreeBuilderCreateNode(tree_builder, node_id));
-
-  std::queue<Node_ID_info<T, L>> cur_level_queue;
-  std::queue<Node_ID_info<T, L>> next_level_queue;
-
-  cur_level_queue.push(Node_ID_info<T, L>(tree_ptr->sparsetree[0], 0));
-  node_id = -1;
-
-  while (!cur_level_queue.empty()) {
-    int cur_level_size = cur_level_queue.size();
-    node_id += cur_level_size;
-
-    for (int i = 0; i < cur_level_size; i++) {
-      Node_ID_info<T, L> q_node = cur_level_queue.front();
-      cur_level_queue.pop();
-
-      bool is_leaf_node = q_node.node.colid == -1;
+tl::Tree<T, T> build_treelite_tree(
+  const DecisionTree::TreeMetaDataNode<T, L> &rf_tree, unsigned int num_class,
+  std::vector<Node_ID_info<T, L>> &cur_level_queue,
+  std::vector<Node_ID_info<T, L>> &next_level_queue) {
+  tl::Tree<T, T> tl_tree;
+  tl_tree.Init();
+
+  // Track head and tail of bounded "queues" (implemented as vectors for
+  // performance)
+  size_t cur_front = 0;
+  size_t cur_end = 0;
+  size_t next_front = 0;
+  size_t next_end = 0;
+
+  cur_level_queue.resize(std::max<size_t>(cur_level_queue.size(), 1));
```

Comment:
I wonder if the whole conversion could become faster if a stack is used instead of a pair of queues. The stack will definitely have a smaller memory footprint (proportional to the tree depth, as opposed to the maximum number of nodes in a single tree level).

This is just a suggestion, however, definitely for another pull request.

Reply:
We need to build in breadth-first traversal order to be consistent with the layout that Treelite formerly enforced with the `CommitModel` method.

Reply:
The breadth-first order is required to create `tl::Tree` because the tree object stores leaf vector elements in a single flat array, and we currently disallow inserting a new leaf vector in the middle of that array. See https://github.com/dmlc/treelite/blob/cf99ff9aefef8c7c99e824fc08662954b95d1de2/include/treelite/tree_impl.h#L633-L654

Reply:
@hcho3 are there any particular node ordering requirements like that inside treelite?

Reply:
@canonizer The ordering requirement exists only when you use the convenience functions `AddChilds()` and `SetLeafVector()`. You can avoid the requirement by manually assigning the leaf vector field. (This will also require a PR on the Treelite side, since the field is currently private.)

Reply:
The same thing applies to the categorical splits as well. With `AddChilds()` and `SetCategoricalSplit()`, you need the breadth-first ordering.

Reply:
I'd considered a Treelite change as well, because doing that would *also* allow us to pre-reserve the space that we need for all nodes. The individual `AllocNode` calls are the remaining bottleneck, but we're fast approaching diminishing returns here. Based on profiling, I think we could get another 2x speedup by making some changes on the Treelite side, but I'd want @JohnZed to weigh in on whether or not that's a priority at this point. Either way, I'd like to defer that to a later PR.

Reply:
And sorry, I should say we could get *at most* a 2x speedup.

Reply:
The categorical split and leaf vector fields are quite annoying, since each split node may contain multiple values for these fields. Previously, I had each split node store a `std::vector`, but that made serialization difficult and time-consuming, so I later consolidated each field into a single vector to store values for all the nodes in that tree.